2.0-proposal/RELATIONS.md

```diff
@@ -0,0 +1,248 @@
+added to the global reactive context variable, where it can be accessed
+by the name of the relation (with dashes changed to underscores).  For example,
+if a charm had the following in its `metadata.yaml`:
+
+```yaml
+provides:
+  apache-website:
+    interface: apache-website
+requires:
+  db:
+    interface: mysql
+peers:
+  peers:
+    interface: myapp
+```
+
+Then the following context attributes would be available:
+
+```python
+from charms.reactive import context
+
+assert context.relations.apache_website
+assert context.relations.db
+assert context.relations.peers
+```
+
+Note that the `apache_website` instance would be of the class that was found in
+the `reactive/relations/apache_website/provides.py` file, the `db` instance
+would be from `reactive/relations/mysql/requires.py`, and the `peers` instance
+would be from `reactive/relations/myapp/peers.py`, in their respective layers.
+
+Also note that those instances would be available even if their relation is not
+related to any remote application.
+
+# The SimpleRelation API
+
+The `SimpleRelation` base class will provide a couple of automatically managed
+reactive flags, as well as providing collections for introspecting the state of
+the relation, including related applications and units, relation data provided
+by those units, and for associating reactive flags with particular related
+applications or units and filtering related applications and units by those
+flags.
+
+## Managed Flags
+
+The automatically managed flags are:
+
+  * `relations.{relation_name}.joined`
+  * `relations.{relation_name}.changed`
+
+where the `{relation_name}` is replaced at runtime with the appropriate
+relation name for that instance.  (The relation name for an instance can also
+be accessed as `relation_instance.relation_name`.)
+
+The `joined` flag will set as long as at least one remote unit is related via
+the relation, and will be updated whenever the list of related units changes so
+that any handler watching the flag will be re-invoked.
+
+The `changed` flag will be set as long as at least one remote unit is
+related and has set data on the relation, and it will be updated whenever
+either a unit updates its relation data or if a related unit goes away.  The
+interface can select only units whose relation data has changed using the
+`with_flag` filter method described below and this flag.
+
+## Application and Unit Collections
+
+The interface layer can iterate over related applications and units using the
+provided collections.  For example:
+
+```python
+class MyRelation(SimpleRelation):
+    class flags:
+        joined = 'relation.{relation_name}.joined'
+
+    @when(flags.joined)
+    def changed(self):
+        for app in self.applications:
+            for unit in app.units:
+                print('Unit {} has data: {}'.format(unit.name, unit.data))
+```
+
+Applications and units can have flags associated with them.  Flags that are
```

Comment:
@galgalesh @johnsca Just because you can, doesn't mean you should :smile: Any channel can be used as a pseudo-(async)-rpc mechanism, provided both ends agree on the protocol.  I was debating about whether complex interactions _between_ charms should be encouraged or not.  Already people ask about "how do I transfer my state flag to the other charm", which reinforces the notion that juju & charms are a kind of distributed programming model (which of course, it _can_ be).

@galgalesh I think you're in favour of increased complexity so that more complex behaviours can be offered _by default_ from charms.reactive? I'm actually in favour of _simpler_ behaviour in charms.reactive, and building out more complex behaviour in a layer (or other library) _if_ that's the route some authors which to go.  I'm _worried_ about having default complex behaviour as it may put off people trying out Juju as an alternative to other configuration management tools.

@johnsca I don't disagree that relations can be used for more complex behaviour.  I would argue that reasoning about that behaviour becomes exponentially more difficult if the data on a relation becomes more dynamic (other than simply changing state due to configuration changes rippling through a set of charms, or leadership changes, etc.).  Otherwise charms become more like 'services' rather than managing services?

@johnsca I also tend to agree on the interfaces; they are effectively 'schema-in-code' in that they enforce the format of the data shared on the boundary between two charms, and also how a charm can access it.  It's an non-versioned schema, but that's an argument for another day. 

My contribution to this debate, possibly defunct, is about how complex it is to write a charm to do the basic job of installing, configuring and managing the lifetime of pieces of software that need to be configured to talk to each other.  More complex behaviours _can_ be built on top of the basic facilities that Juju provides; should this be by default, or as an additional library?